### DISCUSSION TITLE
Secret ability to store Certificates

### DISCUSSION THREAD

**Discussion by wsmelton** [community/community#25521] [Actions]
<p>I am looking at using GitHub Actions to go through a build and deployment process for a module to the PowerShell Gallery. In this process we use a code-signing certificate from an authorized CA. Is there any option/ability for us to save the certificate in GitHub Secrets so we can securely use it within an workflow?</p>
<p>If not, what is a current recommended method?</p>

**Comment by BrightRan** [comment]
<p>Hi <a class="mention" href="/wsmelton">@wsmelton</a> ,</p>
<p>You can try to add the public key of your CA as a secret, and then use it on the workflow to run build and deployment processes.<br>
GitHub automatically redacts secrets printed to the log, replace them with ***, but you should avoid printing secrets to the log intentionally.<br>
More details about secrets, please reference here: <a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/creating-and-using-encrypted-secrets" rel="nofollow noopener">https://help.github.com/en/actions/automating-your-workflow-with-github-actions/creating-and-using-encrypted-secrets</a></p>

**Comment by wsmelton** [comment]
<p>PowerShell we use the command Set-AuthenticodeSignature to sign the files. Our only options in this command is the certificate be in the cert store of the machine or a pfix file.</p>
<p>I would have to write the contents of the secret to a file storing the contents in a secret.</p>

**Comment by timheuer** [comment]
<p>I’m hitting this myself.  Here’s what I did.</p>
<p>Get the base64 encoding of the pfx file and save that as a Secret.  Then I created my own action to extract that and write to a file to later use as an input.  My action: <a href="https://github.com/marketplace/actions/base64-to-file" rel="nofollow noopener">https://github.com/marketplace/actions/base64-to-file</a></p>

**Comment by oldskool** [comment, marked as answer]
<p>Similar to timheuer’s answer, I have many workflows that requires a key file to be present on the runner, I have stored the files contents in a base64 encoded secret and simply decode and write it to a file in my workflow. Like:</p>
<pre><code>- name: Write key file
  env:
    AUTH: ${{ secrets.AUTH }}
  run: |
    echo "$AUTH" | base64 --decode &gt; ${HOME}/key.json
# Authenticate to service...
rm ${HOME/key.json}
</code></pre>
<p>This works perfectly and never exposes the actual key in the workflow output/logs.</p>
